src/k8s/K8sWorkerController.cpp

```diff
@@ -13,66 +13,88 @@ limitations under the License.
+static int TIME_OUT = 900;
+static std::vector<int> activeWorkerIds = {};
+std::mutex workerIdMutex;
+std::mutex k8sSpawnMutex;
+static volatile int nextWorkerId = 0;
+
+static inline int getNextWorkerId(int count) {
+    const std::lock_guard<std::mutex> lock(workerIdMutex);
+    int returnId = nextWorkerId;
+    nextWorkerId += count;
+    return returnId;
+}
+
 K8sWorkerController::K8sWorkerController(std::string masterIp, int numberOfWorkers, SQLiteDBInterface *metadb) {
     this->masterIp = std::move(masterIp);
-    this->numberOfWorkers = numberOfWorkers;
+    this->numberOfWorkers = 0;
+    apiClient_setupGlobalEnv();
     this->interface = new K8sInterface();
     this->metadb = *metadb;
+}
 
-    // Delete all hosts
-    metadb->runUpdate("DELETE FROM host");
+K8sWorkerController::~K8sWorkerController() {
+    delete this->interface;
+    apiClient_unsetupGlobalEnv();
+}
 
-    v1_node_list_t *nodeList = interface->getNodes();
-    listEntry_t *listEntry = NULL;
-    v1_node_t *node = NULL;
+static K8sWorkerController *instance = nullptr;
 
-    std::string query;
-    list_ForEach(listEntry, nodeList->items) {
-        node = static_cast<v1_node_t *>(listEntry->data);
-        query = "INSERT INTO host (name, ip) VALUES ('" + std::string(node->metadata->name) + "', '')";
-        metadb->runInsert(query);
+K8sWorkerController *K8sWorkerController::getInstance() {
+    if (instance == nullptr) {
+        throw std::runtime_error("K8sWorkerController is not instantiated");
     }
+    return instance;
+}
 
-    query = "SELECT idhost, name FROM host";
-    for (auto entry : metadb->runSelect(query)) {
-        nodes[entry.at(1).second] = atoi(entry.at(0).second.c_str());
-    }
+K8sWorkerController *K8sWorkerController::getInstance(std::string masterIp, int numberOfWorkers,
+                                                      SQLiteDBInterface *metadb) {
+    // TODO(thevindu-w): synchronize
+    if (instance == nullptr) {
+        instance = new K8sWorkerController(masterIp, numberOfWorkers, metadb);
+        try {
+            instance->maxWorkers = stoi(instance->interface->getJasmineGraphConfig("max_worker_count"));
+        } catch (std::invalid_argument &e) {
+            controller_logger.error("Invalid max_worker_count value. Defaulted to 4");
+            instance->maxWorkers = 4;
+        }
 
-    // Delete all the workers from the database
-    metadb->runUpdate("DELETE FROM worker");
-    int workersAttached = this->attachExistingWorkers();
-    for (int i = workersAttached; i < numberOfWorkers; i++) {
-        this->spawnWorker(i);
+        // Delete all the workers from the database
+        metadb->runUpdate("DELETE FROM worker");
+        int workersAttached = instance->attachExistingWorkers();
+        if (numberOfWorkers - workersAttached > 0) {
+            instance->scaleUp(numberOfWorkers - workersAttached);
+        }
+    } else {
+        controller_logger.warn("Not initializing again");
```

Comment:
Please add more details to this log message. It does not carry any significant meaning.

Reply:
Removed it